Tests/SkipFoundationTests/Network/TestURL.swift

```diff
@@ -781,6 +781,87 @@ class TestURL : XCTestCase {
        #endif
     }
 
+    func test_appendingPathComponent() {
+        let appendComponent = "foo"
+        let urlsExpected = [
+            ("www.swift.org", "www.swift.org/\(appendComponent)"),
+            ("https://www.swift.org/", "https://www.swift.org/\(appendComponent)"),
+            ("https://www.swift.org/a+b#hash", "https://www.swift.org/a+b/\(appendComponent)#hash"),
+            ("https://www.swift.org/a%20b/#hash?q", "https://www.swift.org/a%20b/\(appendComponent)#hash?q"),
+        ]
+        for (urlString, expected) in urlsExpected {
+            let url = URL(string: urlString)!
+            XCTAssertEqual(url.appendingPathComponent(appendComponent).absoluteString, expected)
+        }
+    }
+
+    func test_appendingPathExtension() {
+        let ext = "foo"
+        let dotExt = ".\(ext)"
+        let urlsExpected = [
+            ("www.swift.org", "www.swift.org\(dotExt)"),
+            ("https://www.swift.org/", "https://www.swift.org/\(dotExt)"),
+            ("https://www.swift.org/a+b#hash", "https://www.swift.org/a+b\(dotExt)#hash"),
+            ("https://www.swift.org/a%20b/c/#hash?q", "https://www.swift.org/a%20b/c\(dotExt)/#hash?q"),
+        ]
+        for (urlString, expected) in urlsExpected {
+            let url = URL(string: urlString)!
+            XCTAssertEqual(url.appendingPathExtension(ext).absoluteString, expected)
+        }
+    }
+
+    func test_deletingLastPathComponent() {
+        let urlsExpected = [ // TODO: support relative paths
+            // ("www.swift.org", "./"),
+            // ("https://www.swift.org/", "https://www.swift.org/../"),
+            ("https://www.swift.org/a", "https://www.swift.org/"),
+            ("https://www.swift.org/a/", "https://www.swift.org/"),
+            ("https://www.swift.org/a/b#hash", "https://www.swift.org/a/#hash"),
+            ("https://www.swift.org/a/b/#hash?q", "https://www.swift.org/a/#hash?q"),
+        ]
+        for (urlString, expected) in urlsExpected {
+            let url = URL(string: urlString)!
+            XCTAssertEqual(url.deletingLastPathComponent().absoluteString, expected)
+        }
+    }
+
+    func test_deletingPathExtension() {
+        let urlsExpected = [
+            ("https://www.swift.org/", "https://www.swift.org/"),
+            ("https://www.swift.org/.", "https://www.swift.org/."),
+            ("https://www.swift.org/a.", "https://www.swift.org/a."),
+            ("https://www.swift.org/a..b", "https://www.swift.org/a."),
+            ("https://www.swift.org/.a.b", "https://www.swift.org/.a"),
+            ("https://www.swift.org/..b", "https://www.swift.org/..b"),
```

Comment:
This fails on macOS 14.5 for me. I believe our CI machines are also an even older version. It also fails on iOS 17. But it passes on iOS 18, and are you on macOS 15? If it's passing for you on macOS 15, then it's clearly a versioning issue. Up to you whether you want to deal with trying to update the test for #if SKIP and if #available or just omit this case

Reply:
Ahh thank you Abe. I did not even consider that these APIs would still be changing.

I'm commenting out the test case since there are doubtless other weird edge cases like this.

Reply:
Looks like the Robolectric CI is passing now, but there is a [new failure](https://github.com/skiptools/skip-foundation/actions/runs/12361630109/job/34499204528#step:16:881) on the Android emulator testing step:

```
12-16 22:04:54.177  2175  2193 I TestRunner: started: testLocalizableStrings$SkipFoundation_debugAndroidTest(skip.foundation.LocaleTests)
12-16 22:04:54.266  2175  2193 E TestRunner: failed: testLocalizableStrings$SkipFoundation_debugAndroidTest(skip.foundation.LocaleTests)
12-16 22:04:54.267  2175  2193 E TestRunner: ----- begin exception -----
12-16 22:04:54.267  2175  2193 E TestRunner: java.lang.NullPointerException
12-16 22:04:54.267  2175  2193 E TestRunner: 	at skip.foundation.URL.deletingLastPathComponent(URL.kt:428)
12-16 22:04:54.267  2175  2193 E TestRunner: 	at skip.foundation.LocaleTests.testLocalizableStrings$SkipFoundation_debugAndroidTest(LocaleTests.kt:305)
12-16 22:04:54.267  2175  2193 E TestRunner: 	at java.lang.reflect.Method.invoke(Native Method)
```

I haven't dug into this, but I suspect the problem may be related to a localization strings URL: the `Localizable.strings` file it is a stored in a traditional Java URL resource. In Robolectric, this will be a plain `file:///` URL (since Robolectric doesn't package things up into an apk), but when deployed in the Android environment, it will be something like `jar:file:/data/app/~~GrNJyKuGMG-gs4i97rlqHg==/skip.ui.test-5w0MhfIK6rNxUpG8yMuXgg==/base.apk!/skip/ui/Resources/Localizable.strings`. I suspect that this might be tripping up some of the new URL logic in this PR.

Ideally you could replicate this by testing against an Android emulator (by setting the `ANDROID_SERIAL=emulator-5554` environment variable, either in Xcode's Scheme run arguments or from the command-line). You could also just paste the URL above into your test case to see if `deletingLastPathComponent()` fails on it.

Reply:
Seems java.net.URI doesn't like the `jar:file:` scheme. It returns null for `rawPath` and `path` (but `toString` still works which is why the old implementation worked (Swift.URL's methods also don't work with these URLs).

I'm not sure what approach to take, I see a couple options:
- Fall back to the old implementation when `rawPath` returns null (should be the least disruptive if this behavior is relied upon elsewhere in the Skip ecosystem, but it's kind of ugly)
- Special case the `jar:file:` URLs in SkipFoundation.URL
- Use the old implementation as custom helpers for processing these URLs in local tests (not sure if other Skip libraries rely on the old behavior of the URL methods?)

Reply:
I don't think jar:file: URLs are entirely uncommon on Android, so that would rule out the last option. 

Do you have any particular strategy in mind for special-casing jar:file: URLs? E.g. trim any "jar:" prefix when performing operations and then add it back?

Reply:
I tried `android.net.Uri` and it returns null paths as well. `java.net.URL` works because it returns `file:...` as part of the path (technically incorrect but enough for these test cases). So potentially swapping to URL is an option? But I don't know what additional side effects that would cause.

Are we sure this is useful as part of the public Skip API? I'm not enough of an Android developer to know how common interacting with raw jar paths is.

Removing the `jar:` prefix is enough for everything to work normally so that's definitely an option. It could be done individually in each method, or the constructor setting a `isJAR` flag and removing the prefix. It doesn't seem ideal to run this logic on every URL to support one special case, but of course I'm deferring to you.

Reply:
> Are we sure this is useful as part of the public Skip API? I'm not enough of an Android developer to know how common interacting with raw jar paths is.

All resources are stored in the jars, so any calls like `Bundle.module.url(forResource: "image", withExtension: "png")` are going to return a `jar:file:` URL.

Reply:
The issue is coming is from the use of `URLComponents`, right? I wonder if there could be some preprocessing and postprocessing of the URL before being passed through that that would remove all but the final `scheme` from the URL before creating the `URLComponents`, and then adding them back when extracting the URL back out from it.

Some thing like this pseudo-code:

```swift
func withURLComponents(f: (URLComponents) -> URI) -> URI {
    var schemes: [String] = []
    var withSingleSchemeURI = self.uri
    if let scheme = withSingleSchemeURI.scheme {
        schemes.append(scheme)
        withSingleSchemeURI = URI(withSingleSchemeURI.substring(scheme.count + 1))
    }
    // add back the final scheme so we have at least one
    if let scheme = schemes.dropLast() {
        withSingleSchemeURI = URI(scheme + ":" + withSingleSchemeURI.absoluteString())
    }

    let components = URLComponents(withSingleSchemeURI)
    var transformed = f(components) // perform the operation in the single-scheme URI URLComponents

    // now add back the deleted schemes
    for scheme in schemes.reversed() {
        transformed = URI(scheme + ":" + transformed.absoluteString())        
    }

    return transformed
}
```

Reply:
> All resources are stored in the jars, so any calls like `Bundle.module.url(forResource: "image", withExtension: "png")` are going to return a `jar:file:` URL.

Ah right.. I'm getting confused by the platform implementations not supporting their own urls but I guess they expect you to be using abstractions.

> The issue is coming is from the use of `URLComponents`, right?

The core of the issue is that `path`/`rawPath` return null for non-standard uris. Which then prevents URLComponents from working.

Considering none of the platform versions support schemes with multiple colons, imho we should special case these `jar:` urls instead of supporting any generic multi-colon scheme. I took a stab at this in the latest commit, though I modified URL rather than URLComponents so that all URL methods will work. But it certainly could be done within URLComponents as well if that's more desirable.

Reply:
Unfortunately, I don't think we can do [this](https://github.com/skiptools/skip-foundation/pull/35/files#diff-70e4edf2e14052846e6fee0c53f315422bb0d1ed3cb0e63c915130a5b8e6c995R132-R135):

```
    public init?(string: String, encodingInvalidCharacters: Bool) {
        let string = string.hasPrefix("jar:file:") ? "jarfile" + string.dropFirst(8) : string
        do {
            self.platformValue = java.net.URI(string) // throws on malformed
```

The problem is that we need the underlying `platformValue` to always be a valid (i.e., loadable) value. For example, Kotlin bridging will be using that underlying value directly when transforming between a `skip.foundation.URL` and a `java.net.URI`, and the scheme `jarfile:` will not be valid. So if Kotlin does `url.platformValue.toURL().openConnection()`, it will fail.

It occurs to me that the easiest solution might be to have [URLComponents](https://github.com/skiptools/skip-foundation/blob/main/Sources/SkipFoundation/URLComponents.swift) tolerate double schemes (like "jar:file:"). I'm not sure where/why it is currently failing, but a test case with a "jar:file:" scheme would probably identify it.

Reply:
Gotcha. I switched to handling this in URLComponents, which means URL params like `path` and `pathComponents` will return empty for these jar urls (exactly the old behavior before this branch). I changed the helper methods to remove their dependency on `platformValue.rawPath` as a result, and added explicit test cases for manipulating these jar urls.

Reply:
Looks good, thanks!

Reply:
Thanks both for all your assistance! Took me a while to get my head around this one.